### ISSUE TITLE
App Rendering Protocol (ARP)

### ISSUE THREAD

**Issue by lite20** [NodeOS/NodeOS#198]
With all the efforts of creating a rendering system, and the discussion on window managers, I'm creating this thread alike so we can discuss how apps will interact with window managers to ultimately render to the display.

Looking at this thread I closed https://github.com/NodeOS/NodeOS/issues/186 I noted that apps need to communicate to the window manager in order to render. This is so window managers have the ultimate power to hide, resize, move and so on and so forth with windows.

If an app has to go through its window manager to render, then the interface needs to be standard among all window managers.

Here's a suggestion of how the interface could work.

Imagining that a window manager listened on a port, it could have a module for apps to interact with it.

The module would expose a Window object. The app create a new instance of this and alter window details such as the title. 

The setting function would be used as so
`window.set("title", "Hello World"); `

And requests to alter anything the window manager didn't support would be ignored so the user could implement anything they wanted from different managers and only what applied to that manager would be honored.

The user could also fetch a canvas object from this window object and this would let apps render to their surface via a canvas-like interface. How this would be implemented is up for discussion.

If we implemented all the methods just as in the actual canvas object this may take a load of code. We could avoid this by making a standard method that takes in a string of the method to be performed and the respective parameters in an array. It's up for discussion.

Behind the scene how the module would forward the commands to the window manager would be up to the designers of the window manager.

**Comment by piranna** [comment]
I like your idea, but I think we should split the API from the protocol.

For the API, I think the obvious solution is to use the [Window object](https://developer.mozilla.org/en-US/docs/Web/API/Window), since it has already all the needed methods to interactuate with a window manager ([resizeTo()](https://developer.mozilla.org/en-US/docs/Web/API/Window/resizeTo), [minimize()](https://developer.mozilla.org/en-US/docs/Web/API/Window/minimize)...) so this way the apps don't need to use a special library/interface to access to NodeOS windows managers and can work with current OSes as independent webapps. This could be easily done by crafting a polyfill library that gets injected when a new app window is created and overwrite the `window` object methods with our own ones, just like Selenium and other similar systems do.

For the protocol we can use Json-RPC both by using a WebSocket or a REST API as you see how it fits. This protocol is what I think we should discuss more, but since there's no need for interactivity and all the operations will be initiated by the user, I think a REST API is more than enough, so we could use [oneshoot](https://github.com/piranna/oneshoot) as basis to use less resources.

Comments? Ideas? :-)


> The user could also fetch a canvas object from this window object and this would let apps render to their surface via a canvas-like interface. How this would be implemented is up for discussion.

If the window managers use iframes for the app windows there's no need for this at all, since they would be the own apps that would decide to use HTML or Canvas or SVG to craft the UI.

**Comment by lite20** [comment]
> Seems to me it's more oriented for the protocol layer, or it should be. I'll review it again to be sure.

I pulled that from here: https://github.com/andersevenrud/OS.js-v2/issues/181

Yes, I was hoping  discuss the protocol. I should habe dubbed it the "ARP, app rendering protocol". But that's beside the matter.

> For the protocol we can use Json-RPC both by using a WebSocket or a REST API as you see how it fits. This protocol is what I think we should discuss more, but since there's no need for interactivity and all the operations will be initiated by the user, I think a REST API is more than enough, so we could use oneshoot as basis to use less resources.

We could and save ports too. But I'd like to introduce another potential interface you guys might like. It's a new notation developed internally at GameFruit for a product were developing. It was designed to streamline the RPC process.

It can be implemented in two ways. I will introduce the simpler way 

Here is our function "joke" which takes in an ID for a joke, and returns its author from a hashmap.

`function lol(joke) {}`

To RPC call this, we could create a JSON object denoting the function we want to call and it's parameters, which would look like this:

`{
    "func":" joke",
    params: ["a bad joke"]
}`

but multiple bytes could be saved if we persay anotated it like this:

`joke/a bad joke`

What happens here is we first put the command to call before the /, and then the parameters after separated by comma's. This can further be simplified by replacing the word joke with a numerical is. If all id's are the same length we don't need a slash to denote where the command name ends.


Why do this? It was originally designed for game servers receiving (litterally) hunreds of thousands of RPC calls per minute. On our server implementation we went the next step in annotating booleans commands and such in binary but that's probably unnecessary. 

Smaller commands are easier to throw around. Their slightly nicer to work with and save bytes. For people with bandwith limits (I cri erry tiem), this is a big bonus. For slower connections, the smaller the data the faster it can be received.

I can easily write a module to work with this in just minutes and maybe I will. It could easily be implemented with oneshoot to create an effective rest API for RPC calls. 

Feedbacks, as always, would be loved! :)

> If the window managers use iframes for the app windows there's no need for this at all, since they would be the own apps that would decide to use HTML or Canvas or SVG to craft the UI.

That sounds excellent. But this does require then that every app create a web server to serve their page? (I count oneshoot as a webserver in this statement)

**Comment by derhuerst** [comment]
@lite20 

> To RPC call this, we could create a JSON object denoting the function we want to call and it's parameters, which would look like this:

I feel like this is reinventing the wheel. What about existing [JS rpc solutions](https://github.com/dominictarr/rpc-stream)?

**Comment by lite20** [comment]
@derhuerst
> I feel like this is reinventing the wheel. What about existing JS rpc solutions?

I don't know it could be. I just suggested it, because it's a system I've used in other products and it works nicely. We could look into other systems. Why don't we start looking and post a few projects that may be eye catching?

In the mean time I'll go create that module after work for anyone who might chose to use it (probably game developers). Should we chose to use it, we'll have a tested module to go off.

**Comment by lite20** [comment]
@derhuerst
> >To RPC call this, we could create a JSON object denoting the function we want to call and it's parameters, which would look like this:

>  feel like this is reinventing the wheel. What about existing JS rpc solutions?

By the way I was just showing the difference between the JSON-RPC calls in existence, and the system I've been working with. We wouldn't actually have to go create these, they're already implemented and in use. :P

**Comment by piranna** [comment]
> To RPC call this, we could create a JSON object denoting the function we want to call and it's parameters, which would look like this:
> 
> I feel like this is reinventing the wheel.

I think so, what @lite20 propose is just [json-rpc](http://www.jsonrpc.org/specification) :-P


> What about existing JS rpc solutions?

I would have done some things different, but definitely a freaking awesome module, thanks! :-D


> By the way I was just showing the difference between the JSON-RPC calls in existence, and the system I've been working with. We wouldn't actually have to go create these, they're already implemented and in use. :P

I would go for standard and well defined formats...

**Comment by piranna** [comment]
> If the window managers use iframes for the app windows there's no need for this at all, since they would be the own apps that would decide to use HTML or Canvas or SVG to craft the UI.
> 
> That sounds excellent. But this does require then that every app create a web server to serve their page? (I count oneshoot as a webserver in this statement)

Yes and no.

From how I see it, we need a web server to serve the static content of the apps (html, client-side Javascript, images...) and after that it's not needed anymore, that's why I created [oneshoot](https://github.com/piranna/oneshoot) and used it with [serve-static](https://github.com/expressjs/serve-static) as example and proof-of-concept. This is somewhat generic and can be used by several apps so they don't need their own server, just the static content.

Following this thinking path, we can see most system requests are related to file operations that could be done by a static web server too, so I pick the previous `oneshoot` and `serve-static` example and improved it to create [Davius](https://github.com/piranna/Davius), a (somewhat) full DAV server that works directly with the filesystem with static files, so you can `GET` web apps static content as usual but also `PUT` files to store them, so apps don't need to use their own server here too.

Being able to get and store static content and files, the only reason why apps would need to use their own server is when they need some interactivity like a terminal or to receive notifications or any other *dynamic* content, that by definition it's somewhat specific to each application and so they require their own server.

So in short: ideally, it's only needed a stateless static web (or WebDAV) server, and a custom server specific for each app that requires dynamic content.

**Comment by lite20** [comment]
> I pick the previous oneshoot and serve-static example and improved it to create Davius, a (somewhat) full DAV server that works directly with the filesystem with static files, so you can GET web apps static content as usual but also PUT files to store them, so apps don't need to use their own server here too.

Epic, and with that I think we're about done here. If anyone has last ccomments or concerns now is a good time. I leave the thread alive a little longer.

**Comment by lite20** [comment]
@piranna you don't happen to have an example use of Davius right? Tryna teach a friend to code at the same time :stuck_out_tongue: Yunitto isn't exactly a "friendly" example.

**Comment by piranna** [comment]
> you don't happen to have an example use of Davius right?

It's included by default in NodeOS latests nighties ;-)

**Comment by lite20** [comment]
The beta of Yunitto has been released on npm. It supports only browser's at the moment. It's a super easy module that'll take complete care of the rendering layer and differences between interfaces. WM Dev's, go ahead and start hacking away! (yes it has been tested. Please report *every* issue on the issues threads)
